File: 0101-0200/0179-Largest Number/0179-Largest Number.cpp
class Solution {
public:
    string largestNumber(vector<int>& nums) {
        if (nums.size() == 0) {
            return "";
        }
        vector<string> strs;
        for (int num : nums) {
            strs.push_back(to_string(num));
        }
        
        sort(strs.begin(), strs.end(), cmp());
        if (strs[0] == "0") {
            return "0";
        }
        string result;
        for (string s : strs) {
            result += s;
        }
        
        return result;
    }

private:
    struct cmp {
        bool operator() (string& s1, string& s2) {
            return s1 + s2 > s2 + s1;
        }  
    };
};

File: 0101-0200/0179-Largest Number/0179-Largest Number.py
from functools import cmp_to_key
class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        def cmp(s1, s2):
            if s1 + s2 > s2 + s1:
                return -1
            elif s1 + s2 < s2 + s1:
                return 1
            else:
                return 0
            
        result = ''.join(sorted(map(str, nums), key=cmp_to_key(cmp)))
        return '0' if result[0] == '0' else result

class Solution2:
    def largestNumber(self, nums: List[int]) -> str:
        class largerNumKey(str):
            def __lt__(x, y):
                return x + y > y + x
            
        result = ''.join(sorted(map(str, nums), key=largerNumKey))
        return '0' if result[0] == '0' else result
